spring-cloud-data-rest/src/main/java/org/springframework/cloud/data/rest/controller/StreamController.java

```diff
@@ -118,10 +125,88 @@ public void save(@RequestParam("name") String name,
 			@RequestParam(value = "deploy", defaultValue = "true")
 			boolean deploy) {
 
-		StreamDefinition definition = new StreamDefinition(name, dsl);
-		this.repository.save(definition);
+		StreamDefinition stream = new StreamDefinition(name, dsl);
+		stream = this.repository.save(stream);
+		if (deploy) {
+			deployStream(stream);
+		}
+	}
+
+	/**
+	 * Request removal of an existing stream definition.
+	 *
+	 * @param name the name of an existing stream definition (required)
+	 */
+	@RequestMapping(value = "/definitions/{name}", method = RequestMethod.DELETE)
+	@ResponseStatus(HttpStatus.OK)
+	public void delete(@PathVariable("name") String name) throws Exception {
+		this.repository.delete(name);
+	}
+
+	/**
+	 * Request removal of all stream definitions.
+	 */
+	@RequestMapping(value = "/definitions", method = RequestMethod.DELETE)
+	@ResponseStatus(HttpStatus.OK)
+	public void deleteAll() throws Exception {
+		this.repository.deleteAll();
+	}
+
+	/**
+	 * Request un-deployment of an existing stream.
+	 *
+	 * @param name the name of an existing stream (required)
+	 */
+	@RequestMapping(value = "/deployments/{name}", method = RequestMethod.DELETE)
+	@ResponseStatus(HttpStatus.OK)
+	public void undeploy(@PathVariable("name") String name) throws Exception {
+		StreamDefinition stream = this.repository.findOne(name);
+		Assert.notNull(stream, String.format("no stream defined: %s", name));
+		undeployStream(stream);
+	}
+
+	/**
+	 * Request un-deployment of all streams.
+	 */
+	@RequestMapping(value = "/deployments", method = RequestMethod.DELETE)
+	@ResponseStatus(HttpStatus.OK)
+	public void undeployAll() throws Exception {
+		for (StreamDefinition stream : this.repository.findAll()) {
+			this.undeployStream(stream);
+		}
+	}
+
+	/**
+	 * Request deployment of an existing stream definition. The name must be included in the path.
+	 *
+	 * @param name the name of an existing stream definition (required)
+	 * @param properties the deployment properties for the stream as a comma-delimited list of key=value pairs
+	 */
+	@RequestMapping(value = "/deployments/{name}", method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.CREATED)
+	public void deploy(@PathVariable("name") String name, @RequestParam(required = false) String properties) throws Exception {
+		StreamDefinition stream = this.repository.findOne(name);
+		Assert.notNull(stream, String.format("no stream defined: %s", name));
+		deployStream(stream);
+	}
 
-		// todo: deploy
+	private void deployStream(StreamDefinition stream) {
+		Iterator<ModuleDefinition> iterator = stream.getDeploymentOrderIterator();
+		for (int i = 0; iterator.hasNext(); i++) {
+			ModuleDefinition module = iterator.next();
+			// todo: support processor
```

Comment:
that would work I think... probably change "sinkIndex" to "sourceIndex" (they're in reverse order here)

we can revisit after this is merged though since at the moment the registry itself only supports `time-source` and `log-sink` ;)